### PR TITLE
Conform SAML2INT; check AuthnReqsignature unless specifically disabled

### PR DESCRIPTION
**Pull request by tvdijen**
Closes https://github.com/simplesamlphp/simplesamlphp/issues/1407

### CODE REVIEW

**Review by jaimeperez — APPROVED** [review]

**Comment by tvdijen** [comment]
Needs a field-test as per linked issue to confirm this fixes the compliancy issue

**Review by thijskh — CHANGES_REQUESTED** [review]
What I'm missing is a docs change... this behaviour must really be in the documentation of these options I think.

**Comment by tvdijen** [comment]
I do not disagree, but we in general say to comply SAML2INT, right?

**Comment by tvdijen** [comment]
Field-test is done; hence the fix.
I'd rather not use the global variable, but at this point I see no other way..

**Comment by thijskh** [comment]
> I do not disagree, but we in general say to comply SAML2INT, right?

It is sure to break existing configs. So while this is not a problem per se, it is nice if people can then quickly find what changed so they can fix it.

File: modules/saml/lib/Message.php [hunk 1]
```diff
@@ -213,7 +213,10 @@ public static function validateMessage(
             }
         }
 
-        if ($enabled === null) {
+        // If not specifically set to false, the signature must be checked to conform to SAML2INT
+        if (isset($_REQUEST['Signature']) && ($enabled !== false)) {
```
Comment (thijskh) [inline modules/saml/lib/Message.php:217]: This does not work with the POST binding, does it?

File: docs/simplesamlphp-reference-sp-remote.md [hunk 2]
```diff
@@ -300,6 +300,9 @@ The following options can be set:
 :   Note that this option also exists in the IdP-hosted metadata.
     The value in the SP-remote metadata overrides the value in the IdP-hosted metadata.
 
+:   Note: if unset or `true`. this setting will enforce a valid signature.
```
Comment (thijskh) [inline docs/simplesamlphp-reference-sp-remote.md:303]: I don’t think this correcty describes the new behaviour. It is now this:

- true: authnrequest must be signed (and sig will be validated)
- null: authnrequest may be signed, if it is, sig will be validated
- false: authnrequest sig is never checked

File: modules/saml/lib/Message.php [hunk 3]
```diff
@@ -221,9 +224,7 @@ public static function validateMessage(
 
         if (!$enabled) {
             return;
-        }
-
-        if (!self::checkSign($srcMetadata, $message)) {
+        } elseif (!self::checkSign($srcMetadata, $message)) {
```
Comment (thijskh) [inline modules/saml/lib/Message.php:228]: This change does not seem to do anything, why is it changed?

**Review by thijskh — CHANGES_REQUESTED** [review]

File: modules/saml/lib/Message.php (hunk 3 above)
Comment (tvdijen) [reply modules/saml/lib/Message.php:232]: For readability... No actual change

File: docs/simplesamlphp-reference-sp-remote.md (hunk 2 above)
Comment (tvdijen) [reply docs/simplesamlphp-reference-sp-remote.md]: You are right.. I've updated it.. Better?

File: modules/saml/lib/Message.php (hunk 1 above)
Comment (tvdijen) [reply modules/saml/lib/Message.php]: Good catch! Fixed it

File: docs/simplesamlphp-reference-sp-remote.md (hunk 2 above)
Comment (thijskh) [reply docs/simplesamlphp-reference-sp-remote.md]: I would just make the three values and their meanings explicit in the docs for idp-hosted/sp-remote, like my suggestition, so people don't have to combine different fragments to conclude what does what.

**Pull request merged**